org.eclipse.xtext.xbase.testing/src/org/eclipse/xtext/xbase/testing/InMemoryJavaCompiler.java

```diff
@@ -0,0 +1,287 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.xbase.testing;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.core.compiler.CategorizedProblem;
+import org.eclipse.jdt.core.compiler.CharOperation;
+import org.eclipse.jdt.internal.compiler.ClassFile;
+import org.eclipse.jdt.internal.compiler.CompilationResult;
+import org.eclipse.jdt.internal.compiler.DefaultErrorHandlingPolicies;
+import org.eclipse.jdt.internal.compiler.ICompilerRequestor;
+import org.eclipse.jdt.internal.compiler.batch.CompilationUnit;
+import org.eclipse.jdt.internal.compiler.classfmt.ClassFileReader;
+import org.eclipse.jdt.internal.compiler.classfmt.ClassFormatException;
+import org.eclipse.jdt.internal.compiler.env.ICompilationUnit;
+import org.eclipse.jdt.internal.compiler.env.INameEnvironment;
+import org.eclipse.jdt.internal.compiler.env.NameEnvironmentAnswer;
+import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
+import org.eclipse.jdt.internal.compiler.problem.DefaultProblemFactory;
+import org.eclipse.xtext.util.JavaVersion;
+import org.eclipse.xtext.xbase.lib.Conversions;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.ListExtensions;
+
+/**
+ * @since 2.9
+ */
+public class InMemoryJavaCompiler {
+	private static class ClassLoaderBasedNameEnvironment implements INameEnvironment {
+		private final ClassLoader classLoader;
+
+		private Map<String, NameEnvironmentAnswer> cache = new HashMap<>();
+
+		@Override
+		public void cleanup() {
+			cache.clear();
+		}
+
+		@Override
+		public NameEnvironmentAnswer findType(char[][] compoundTypeName) {
+			try {
+				String fileName = new String(CharOperation.concatWith(compoundTypeName, '/')) + ".class";
+				if (cache.containsKey(fileName)) {
+					return cache.get(fileName);
+				}
+				URL url = classLoader.getResource(fileName);
+				if (url == null) {
+					cache.put(fileName, null);
+					return null;
+				}
+				ClassFileReader reader = ClassFileReader.read(url.openStream(), fileName);
```

Comment:
use try-with-resources for url.openStream()?

Reply:
please provide a proposal.
have no head for this today

Reply:
```
				try (InputStream is = url.openStream()) {
					ClassFileReader reader = ClassFileReader.read(is, fileName);
					NameEnvironmentAnswer result = new NameEnvironmentAnswer(reader, null);
					cache.put(fileName, result);
					return result;
				}
```